move-vm-backend-common/src/move_bytecode.rs

```diff
@@ -0,0 +1,70 @@
+//! Tools for bytecode verification.
+
+use move_binary_format::file_format::{CompiledScript, Signature, SignatureToken};
+use move_core_types::vm_status::StatusCode;
+
+/// Check the script bytecode and ensure the signer rule is followed.
+/// Upon success, return the number of signers in the script argument list.
+///
+/// The signer rule:
+/// A transaction script can have an arbitrary number of signers as long as the signers are a prefix to
+/// any other arguments. In other words, all of the signer arguments must come first:
+/// ```move
+/// script {
+///     use std::signer;
+///     fun main(s1: signer, s2: signer, x: u64, y: u8) {
+///         // ...
+///     }
+/// }
+/// ```
+pub fn verify_script_integrity_and_check_signers(script_bc: &[u8]) -> Result<usize, StatusCode> {
+    let compiled_script = CompiledScript::deserialize(script_bc).map_err(|e| e.major_status())?;
+
+    let Signature(script_params) = &compiled_script
+        .signatures
+        .first()
+        .ok_or(StatusCode::INVALID_SIGNATURE)?;
+
+    let mut signer_param_cnt = 0;
+    // Find all signer params at the beginning of the parameter list.
+    for ty in script_params.iter() {
+        // Allow pure signers here - not even references to signers.
+        if !ty.is_signer() {
+            break;
+        }
+
+        signer_param_cnt += 1;
+    }
+
+    // Check that the rest of the parameter list contains no hidden signers or unallowed types.
+    for ty in script_params[signer_param_cnt..].iter() {
+        let is_script_param_type_allowed = is_valid_txn_arg(ty);
+
+        if !is_script_param_type_allowed {
+            return Err(StatusCode::INVALID_MAIN_FUNCTION_SIGNATURE);
+        }
+    }
+
+    Ok(signer_param_cnt)
+}
+
+// Return whether the argument is valid/allowed and whether it needs construction.
```

Comment:
Shall this method return whether a parameter:
- is valid XOR allowed?
- is valid AND allowed?
- is valid OR allowed?
- or does valid in this context mean allowed?

It's only about the description "valid/allowed"... maybe we just remove "valid" and write only "allowed"?

Reply:
Done